js/meshopt_clusterizer.js

```diff
@@ -0,0 +1,380 @@
+
+			// copy input data to wasm memory
+			heap = new Uint8Array(instance.exports.memory.buffer);
+			heap.set(bytes(m.vertices), meshlet_verticesp);
+			heap.set(bytes(m.triangles), meshlet_trianglesp);
+
+			instance.exports.meshopt_computeMeshletBounds(
+				resultp,
+				meshlet_verticesp,
+				meshlet_trianglesp,
+				m.triangles.length / 3,
+				verticesp,
+				vertex_count,
+				vertex_positions_stride
+			);
+
+			// heap might (will?) have grown -> re-acquire
+			heap = new Uint8Array(instance.exports.memory.buffer);
+
+			var boundsBytes = heap.subarray(resultp, resultp + BOUNDS_SIZE);
+
+			results.push({
+				center: copyTypedArray(new Float32Array(boundsBytes.buffer, boundsBytes.byteOffset, 3)),
+				radius: new Float32Array(boundsBytes.buffer, boundsBytes.byteOffset + 3 * Float32Array.BYTES_PER_ELEMENT, 1)[0],
+				coneApex: copyTypedArray(new Float32Array(boundsBytes.buffer, boundsBytes.byteOffset + 4 * Float32Array.BYTES_PER_ELEMENT, 3)),
+				coneAxis: copyTypedArray(new Float32Array(boundsBytes.buffer, boundsBytes.byteOffset + 7 * Float32Array.BYTES_PER_ELEMENT, 3)),
+				coneCutoff: new Float32Array(boundsBytes.buffer, boundsBytes.byteOffset + 10 * Float32Array.BYTES_PER_ELEMENT, 1)[0],
+				coneAxisS8: copyTypedArray(new Int8Array(boundsBytes.buffer, boundsBytes.byteOffset + 11 * Float32Array.BYTES_PER_ELEMENT, 3)),
+				coneCutoffS8: new Int8Array(boundsBytes.buffer, boundsBytes.byteOffset + 47, 1)[0],
+			});
+
+			// reset meshlet memory
+			sbrk(meshlet_verticesp - sbrk(0));
+		}
+
+		// reset memory
+		sbrk(verticesp - sbrk(0));
+
+		return results.length > 1 ? results : results[0];
+	}
+
+	function computeClusterBounds(indices, vertex_positions, vertex_positions_stride) {
+		assert(vertex_positions.byteLength % vertex_positions_stride === 0);
+		var vertex_count = vertex_positions.byteLength / vertex_positions_stride;
+
+		var sbrk = instance.exports.sbrk;
+
+		// allocate memory
+		var resultp = sbrk(BOUNDS_SIZE);
+		var indicesp = sbrk(indices.byteLength);
+		var verticesp = sbrk(vertex_positions.byteLength);
+
+		// copy input data to wasm memory
+		var heap = new Uint8Array(instance.exports.memory.buffer);
+		heap.set(bytes(indices), indicesp);
+		heap.set(bytes(vertex_positions), verticesp);
+
+		instance.exports.meshopt_computeClusterBounds(
+			resultp,
+			indicesp,
+			indices.length,
+			verticesp,
+			vertex_count,
+			vertex_positions_stride
+		);
+
+		// heap might (will?) have grown -> re-acquire
+		heap = new Uint8Array(instance.exports.memory.buffer);
+
+		var boundsBytes = heap.subarray(resultp, resultp + BOUNDS_SIZE);
+
+		var result = {
+			center: copyTypedArray(new Float32Array(boundsBytes.buffer, boundsBytes.byteOffset, 3)),
+			radius: new Float32Array(boundsBytes.buffer, boundsBytes.byteOffset + 3 * Float32Array.BYTES_PER_ELEMENT, 1)[0],
+			coneApex: copyTypedArray(new Float32Array(boundsBytes.buffer, boundsBytes.byteOffset + 4 * Float32Array.BYTES_PER_ELEMENT, 3)),
```

Comment:
What does copyTypedArray do here and elsewhere?

Reply:
The same as `TypedArray.prototype.slice` - removed it

Reply:
Ah sorry, that doesn't really answer the question. It copies the data from a typed array to new typed array of the same type. The `new Float32Array(boundsBytes.buffer, ...)` is more like a cast where the new float array is backed by the same memory as `boundsBuffer` and thus the wasm heap. Instead, the `copyToTypedArray` / `TypedArray.prototype.slice` allocates a new chunk of memory and copies the data from the wasm heap so that it isn't overridden by the next wasm call or gets detached when the heap grows (or shrinks? Does that happen or is it "once grown always grown"?). Maybe there's a more elegant solution. I haven't worked with transforming typed arrays a lot and find the api confusing at times. Let me know if we should change this.